### PR TITLE
Update readme for the 3.0.0 release

### PR DESCRIPTION
**Pull request by petea**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -228,26 +224,37 @@ GTMSessionFetcher *fetcher = [fetcherService fetcherWithURL:userinfoEndpoint];
...
+// Create a GIDKeychainStore instance
+GIDKeychainStore *keychainStore =
+    [[GIDKeychainStore alloc] initWithItemName:kGTMAppAuthExampleAuthorizerKey];
+    
+NSError *error;
+
 // Save to Keychain
-[GTMAppAuthFetcherAuthorization saveAuthorization:_authorization
-                                toKeychainForName:kGTMAppAuthExampleAuthorizerKey];
+[keychainStore saveWithAuthSession:self.authSession error:&error];
```
Comment (petea) [inline README.md:237]: @mdmathias we might want to rename the method`saveWithAuth...` -> `saveAuth...`.  Thoughts?

Comment (mdmathias) [reply README.md:237]: @petea agreed. [Changed here](https://github.com/google/GTMAppAuth/pull/207/commits/68f33f07cf37f4904b6e728f2c029b30131df85b).

File: README.md [hunk 2]
```diff
@@ -395,24 +406,28 @@ GTMOAuth2 keychain, and the new details for all other GTMAppAuth calls.
...
-  if (authorization) {
+  GTMKeychainStore oldKeychainStore = [[GTMKeychainStore alloc] initWithItemName:kPreviousKeychainName];
+  authSession =
+      [oldKeychainStore retrieveAuthSessionInGTMOAuth2FormatWithClientID:kPreviousClientID
+                                                            clientSecret:kPreviousClientSecret
+                                                            error:&error];
+  if (authSession) {
     // Remove previously stored GTMOAuth2-formatted data.
-    [GTMOAuth2KeychainCompatibility removeAuthFromKeychainForName:kPreviousKeychainName];
+    [oldKeychainStore removeAuthSessionAndReturnError:&error];
```
Comment (petea) [inline README.md:428]: @mdmathias should this be `removeAuthSessionWithError:`?

File: README.md [hunk 3]
```diff
@@ -228,26 +224,37 @@ GTMSessionFetcher *fetcher = [fetcherService fetcherWithURL:userinfoEndpoint];
...
+[keychainStore saveAuthSession:self.authSession error:&error];
+if (error) {
+  // Handle error
+}
 
-// Restore from Keychain
-GTMAppAuthFetcherAuthorization* authorization =
-    [GTMAppAuthFetcherAuthorization authorizationFromKeychainForName:kGTMAppAuthExampleAuthorizerKey];
+// Retrieve from Keychain
+self.authSession = [keychainStore retrieveAuthSessionAndReturnError:&error];
```
Comment (petea) [inline README.md:243]: @mdmathias should this be `retrieveAuthSessionWithError:`?

File: README.md (hunk 2 above)
Comment (mdmathias) [reply README.md:428]: https://github.com/google/GTMAppAuth/pull/216

File: README.md (hunk 3 above)
Comment (mdmathias) [reply README.md:243]: https://github.com/google/GTMAppAuth/pull/216

File: README.md [hunk 4]
```diff
@@ -228,26 +224,37 @@ GTMSessionFetcher *fetcher = [fetcherService fetcherWithURL:userinfoEndpoint];
 
 ### Saving to the Keychain
 
-You can easily save `GTMAppAuthFetcherAuthorization` instances to the Keychain using
-the included `GTMAppAuthFetcherAuthorization+Keychain` category.
+You can easily save `GTMAuthSession` instances to the Keychain using the `GTMKeychainStore` class.
```
Comment (mdmathias) [inline README.md:227]: Is it worth mentioning that we provide that as a convenience? I.e., that they can make their own conformance to `AuthSessionStore` to persist `AuthSession`s.

File: README.md [hunk 5]
```diff
@@ -228,26 +224,37 @@ GTMSessionFetcher *fetcher = [fetcherService fetcherWithURL:userinfoEndpoint];
...
+// Create a GIDKeychainStore instance
+GIDKeychainStore *keychainStore =
+    [[GIDKeychainStore alloc] initWithItemName:kGTMAppAuthExampleAuthorizerKey];
+    
+NSError *error;
+
 // Save to Keychain
-[GTMAppAuthFetcherAuthorization saveAuthorization:_authorization
-                                toKeychainForName:kGTMAppAuthExampleAuthorizerKey];
+[keychainStore saveAuthSession:self.authSession error:&error];
```
Comment (mdmathias) [inline README.md:237]: Worth adding a comment that says the `authSession` is saved with the key given during the `KeychainStore`'s initialization?

File: README.md [hunk 6]
```diff
@@ -228,26 +224,37 @@ GTMSessionFetcher *fetcher = [fetcherService fetcherWithURL:userinfoEndpoint];
...
+if (error) {
+  // Handle error
+}
 
-// Restore from Keychain
-GTMAppAuthFetcherAuthorization* authorization =
-    [GTMAppAuthFetcherAuthorization authorizationFromKeychainForName:kGTMAppAuthExampleAuthorizerKey];
+// Retrieve from Keychain
```
Comment (mdmathias) [inline README.md:242]: `// Retrieve from Keychain with kGTMAppAuthExampleAuthorizerKey given during initialization`?

File: README.md [hunk 7]
```diff
@@ -260,41 +267,43 @@ key is set to
...
+GTMKeychainStore keychainStore = [[GTMKeychainStore alloc] initWithItemName:kKeychainItemName];
 
-// Remove from Keychain
-[GTMOAuth2KeychainCompatibility removeAuthFromKeychainForName:kKeychainItemName];
+// Retrieve from the Keychain
+NSError *error;
+GTMAuthSession *authSession =
+    [keychainStore retrieveAuthSessionForGoogleInGTMOAuth2FormatWithClientID:clientID
+                                                                clientSecret:clientSecret
+                                                                error:&error];
```
Comment (mdmathias) [inline README.md:295]: Formatting: align `error:` with the other parameters.

File: README.md [hunk 8]
```diff
@@ -395,24 +406,28 @@ GTMOAuth2 keychain, and the new details for all other GTMAppAuth calls.
...
-  authorization =
-      [GTMOAuth2KeychainCompatibility authForGoogleFromKeychainForName:kPreviousKeychainName
-                                                              clientID:kPreviousClientID
-                                                          clientSecret:kPreviousClientSecret];
-  if (authorization) {
+  GTMKeychainStore oldKeychainStore = [[GTMKeychainStore alloc] initWithItemName:kPreviousKeychainName];
+  authSession =
+      [oldKeychainStore retrieveAuthSessionInGTMOAuth2FormatWithClientID:kPreviousClientID
+                                                            clientSecret:kPreviousClientSecret
+                                                            error:&error];
```
Comment (mdmathias) [inline README.md:425]: Align `error:` with other parameters.

**Review by mdmathias — CHANGES_REQUESTED** [review]
Is it worth adding a session covering the new error (`KeychainStore.Error`) and how it impacts error handling?

File: README.md (hunk 4 above)
Comment (petea) [reply README.md:227]: Added an "Implementing Your Own Storage" section below.

File: README.md (hunk 5 above)
Comment (petea) [reply README.md:239]: Added an explanation in the initialization comment above.

File: README.md (hunk 6 above)
Comment (petea) [reply README.md:244]: See above.

File: README.md (hunk 7 above)
Comment (petea) [reply README.md]: done

File: README.md (hunk 8 above)
Comment (petea) [reply README.md]: done

File: README.md [hunk 9]
```diff
@@ -260,41 +269,49 @@ key is set to
...
-Keychain methods are provided in `GTMOAuth2KeychainCompatibility`.
+To assist the migration from GTMOAuth2 to GTMAppAuth, GTMOAuth2-compatible Keychain methods are provided in `GTMKeychainStore`.
 
 ```objc
-// Restore from Keychain
-GTMAppAuthFetcherAuthorization *auth =
-    [GTMOAuth2KeychainCompatibility authForGoogleFromKeychainForName:kKeychainItemName
-                                                            clientID:clientID
-                                                        clientSecret:clientSecret];
+GTMKeychainStore keychainStore = [[GTMKeychainStore alloc] initWithItemName:kKeychainItemName];
```
Comment (mdmathias) [inline README.md:296]: Above we use the key name `kGTMAppAuthExampleAuthorizerKey`. Worth simplifying to using just one?

File: README.md [hunk 10]
```diff
@@ -370,16 +384,21 @@ The client ID used for GTMAppAuth is [different](#oauth-client-registration) to
...
-Change how you serialize your `authorization` object using the new methods
-using the following example.
+Change how you serialize your `authorization` object by using `GTMAuthSession` and `GTMKeychainStore` as follows:
 
 ```objc
+// Create an auth session from AppAuth's auth state object
+GTMAuthSession *authSession = [[GTMAuthSession alloc] initWithAuthState:authState];
+
+// Create a keychain store
+GTMKeychainStore keychainStore = [[GTMKeychainStore alloc] initWithItemName:kNewKeychainName];
```
Comment (mdmathias) [inline README.md:397]: Same note re: key name.

File: README.md [hunk 11]
```diff
@@ -395,24 +414,28 @@ GTMOAuth2 keychain, and the new details for all other GTMAppAuth calls.
 Keychain migration example:
 
 ```objc
+// Create a keychain store
+GTMKeychainStore keychainStore = [[GTMKeychainStore alloc] initWithItemName:kNewKeychainName];
```
Comment (mdmathias) [inline README.md:418]: Ditto

**Review by mdmathias — APPROVED** [review]
Just a small question about example keychain names in this documentation.

File: README.md (hunk 9 above)
Comment (petea) [reply README.md:294]: Went with `k[New]KeychainItemName`.

**Pull request merged**